src/plugins/upf/dpi.c

```diff
@@ -0,0 +1,1123 @@
+
+          vlib_cli_output (vm, "regex: %s, app: %s", *regex, app->name);
+        }
+    }
+  else
+    {
+      error = clib_error_return (0, "DB does not exist...");
+    }
+
+done:
+  vec_free (name);
+  unformat_free (line_input);
+
+  return error;
+}
+
+/* *INDENT-OFF* */
+VLIB_CLI_COMMAND (upf_dpi_show_db_command, static) =
+{
+  .path = "show upf dpi app",
+  .short_help = "show upf dpi app <name>",
+  .function = upf_dpi_show_db_command_fn,
+};
+/* *INDENT-ON* */
+
+/* Action function shared between message handler and debug CLI */
+
+static int
+vnet_upf_rule_add_del(u8 * app_name, u32 rule_index, u8 add,
+                      upf_rule_args_t * args);
+
+static int
+vnet_upf_app_add_del(u8 * name, u8 add);
+
+int upf_app_add_del (upf_main_t * sm, u8 * name, int add)
+{
+  int rv = 0;
+
+  rv = vnet_upf_app_add_del(name, add);
+
+  return rv;
+}
+
+int upf_rule_add_del (upf_main_t * sm, u8 * name, u32 id,
+                      int add, upf_rule_args_t * args)
+{
+  int rv = 0;
+
+  rv = vnet_upf_rule_add_del(name, id, add, args);
+
+  return rv;
+}
+
+static int
+vnet_upf_app_add_del(u8 * name, u8 add)
+{
+  upf_main_t *sm = &upf_main;
+  upf_dpi_app_t *app = NULL;
+  u32 index = 0;
+  u32 rule_index = 0;
+  uword *p = NULL;
+
+  p = hash_get_mem (sm->upf_app_by_name, name);
+
+  if (add)
+    {
+      if (p)
+        return VNET_API_ERROR_VALUE_EXIST;
+
+      pool_get (sm->upf_apps, app);
+      memset(app, 0, sizeof(*app));
+
+      app->name = vec_dup(name);
+      app->rules_by_id = hash_create_mem (0, sizeof (u32), sizeof (uword));
+      app->path_db_index = ~0;
+      app->host_db_index = ~0;
+
+      hash_set_mem (sm->upf_app_by_name, app->name, app - sm->upf_apps);
+    }
+  else
+    {
+      if (!p)
+        return VNET_API_ERROR_NO_SUCH_ENTRY;
+
+      hash_unset_mem (sm->upf_app_by_name, name);
+      app = pool_elt_at_index (sm->upf_apps, p[0]);
+
+      /* *INDENT-OFF* */
+      hash_foreach(rule_index, index, app->rules_by_id,
+      ({
+         upf_dpi_rule_t *rule = NULL;
+         rule = pool_elt_at_index(app->rules, index);
+         vnet_upf_rule_add_del(app->name, rule->id, 0, NULL);
+      }));
+      /* *INDENT-ON* */
+
+      upf_dpi_remove(app->path_db_index);
+      upf_dpi_remove(app->host_db_index);
+      vec_free (app->name);
+      hash_free(app->rules_by_id);
+      pool_free(app->rules);
+      pool_put (sm->upf_apps, app);
+    }
+
+  return 0;
+}
+
+static clib_error_t *
+upf_create_app_command_fn (vlib_main_t * vm,
+                           unformat_input_t * input,
+                           vlib_cli_command_t * cmd)
+{
+  unformat_input_t _line_input, *line_input = &_line_input;
+  u8 *name = NULL;
+  clib_error_t *error = NULL;
+  int rv = 0;
+
+  /* Get a line of input. */
+  if (!unformat_user (input, unformat_line_input, line_input))
+    return error;
+
+  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "%s", &name))
```

Comment:
```unformat (line_input, "%v", &name)``` would convert it into a vector without the zero termination (see also https://github.com/travelping/vpp/pull/5/files#r220119212)

Reply:
Update, VPP uses %v to scan strings into vectors (without the terminating 0), it uses %s to scan into the typical zero terminated C strings.
Please replace all place where it make sense where you use %s in unformat with %v.